src/include/ipc/dai.h

```diff
@@ -74,16 +74,37 @@ struct sof_ipc_dai_config {
 	uint16_t reserved16;	/**< alignment */
 
 	/* reserved for future use */
-	uint32_t reserved[8];
+	uint32_t reserved[4];
+} __attribute__((packed));
+
+struct sof_ipc_multidai_params {
+	struct sof_ipc_cmd_hdr hdr;
+
+	uint32_t num_dais;
+	uint32_t reserved[5];
+
+	struct sof_ipc_stream_map stream_map;
+	struct sof_ipc_dai_base_params params[0];
```

Comment:
I am a bit lost here.

The sof_ipc_dai_config already contains a 'sof_ipc_dai_base_params'.

So in the multi-dai case, is the baseline ignored?

Also in the size of this array determined by 'num_dais'?

Reply:
The base params inside of sof_ipc_dai_config is for the DAI you're directly communicating with.
The base params inside of the multidai is for the children of the multidai. And yes, the size is determined by num_dais.

To explain a bit more what's going on here:
The union of arrays contained within sof_ipc_dai_config contain structures, which don't convey basic DAI information (dai type and index). So those need to be passed separately. That's the reason behind the additional base params array.

Reply:
let's take the example of a multi-dai SounDwire/ALH case.

The dai_config would contain
hdr
base_params (multi_DAI type + index)

So then the multidai array should contain the ALH information for each child DAI, right?

The problem is that I don't see anything ALH-related in the multi_dai_params...only a stream_map and a base_params (which can only mean individual DAI and type). Where would the ALH stream_id passed?

Reply:
As I understood from the previous discussion, ALH related params would be in the flex array of the parent struct, which is in sof_ipc_dai_config.

To be specific: the first element of that flex array in this case would be the sof_ipc_dai_multidai_params. And directly after that, the very next members would be the children for multida, namely sof_ipc_dai_alh_params.

Reply:
sorry, I don't get how this would work.
You don't have an array of unions, you have a union of arrays.

I don't think you can make the type of structure vary between array elements?

Reply:
I'm pretty sure you can, since from the standpoint of the compiler, all union members have a size of 0 anyway. So, the union is purely cosmetic, serving as a template for the code. Now that I think about this, the union doesn't do anything at all here, might as well get rid of it.

Reply:
Honestly, the more I think about this, the more I realize that there's no good way of actually having multiple configs for different DAI types in a single IPC.

- As you mentioned, having a union of flex arrays suggests we can only have 1 type, doing otherwise is indeed bad practice.
- Having a flex array of unions has padding issues I mentioned earlier.
- Having multiple flex arrays, without a union also doesn't smell right

I'd honestly abandon those efforts at this point and revert to the initial proposal: not touching the DAI config, simply adding a new DAI config member with its own union of flex arrays, limited to a single type.

Reply:
@lgirdwood @plbossart can we come to an agreement on this one?

Reply:
@lgirdwood @plbossart can we come to an agreement on this one?

Reply:
@slawblauciak is the sticking point the unions or the ALH ? I defer to @plbossart for all ALH and Soundwire stuff.

Reply:
It's the unions. I don't think it's sane to try to pack all the different configurations into a single IPC  inside of sof_ipc_dai_config. As I mentioned, I'd rather do this by the book and just do as in the initial proposal: a new union member in sof_ipc_dai_config, one that would have its own, similar union, except of flex arrays.